Userland/Libraries/LibGfx/MedianCut.cpp

```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright (c) 2024, Lucas Chollet <lucas.chollet@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/QuickSort.h>
+#include <AK/Statistics.h>
+#include <LibGfx/MedianCut.h>
+
+namespace Gfx {
+
+namespace {
+
+using Bucket = Vector<ARGB32>;
+using Buckets = Vector<Bucket>;
+
+void sort_along_color(Bucket& bucket, u8 color_index)
+{
+    auto less_than = [=](ARGB32 first, ARGB32 second) {
+        auto const first_color = Color::from_argb(first);
+        auto const second_color = Color::from_argb(second);
+        switch (color_index) {
+        case 0:
+            return first_color.red() < second_color.red();
+        case 1:
+            return first_color.green() < second_color.green();
+        case 2:
+            return first_color.blue() < second_color.blue();
+        default:
+            VERIFY_NOT_REACHED();
+        }
+    };
+
+    AK::quick_sort(bucket, less_than);
+}
+
+template<typename T>
+struct MaxAndIndex {
+    T maximum;
+    u32 index;
+};
+
+template<typename T, class MoreThan>
+MaxAndIndex<T> max_and_index(Span<T> values, MoreThan more_than)
+{
+    VERIFY(values.size() != 0);
+
+    u32 max_index = 0;
+    RemoveCV<T> max_value = values[0];
+    for (u32 i = 0; i < values.size(); ++i) {
+        if (more_than(values[i], max_value)) {
+            max_value = values[i];
+            max_index = i;
+        }
+    }
+
+    return { max_value, max_index };
+}
+
+ErrorOr<void> split_bucket(Buckets& buckets, u32 index_to_split, u8 color_index)
+{
+    auto& to_split = buckets[index_to_split];
+
+    sort_along_color(to_split, color_index);
+
+    Bucket new_bucket {};
+
+    auto const middle = ceil_div(to_split.size(), 2ul);
+
+    auto const span_to_move = to_split.span().slice(middle);
+    for (u32 i = 0; i < span_to_move.size(); ++i)
+        TRY(new_bucket.try_append(span_to_move[i]));
```

Comment:
vector has a span ctor. Does `Bucket new_bucket { span_to_move };` work?

Reply:
We should rather make `extend` accept a `Span`

Reply:
Sure, but here the thing starts out empty anyway, no?

Reply:
Yes, but I want to keep that code OOM safe and constructors are not.

Reply:
I left a FIXME about it.